### PR TITLE
Allowing to compute lambda fitting on series and stop dropping PM if it is in the REE pattern

### PR DESCRIPTION
**Pull request by mmuesly**
## Description
We love the implementation for characterizing tetrades in REE patterns on a dataframe. For performance considerations, we require to wrap the implementation into an apply function on a dataframe. In its current form, it is not possible to compute lambdas only on a single series. I suggest to change the implementation to also support the core computation of lambdas on series objects. As a consequence, it can be used inside an apply function and offers easy parallelization in combination with the [pandarallel](https://github.com/nalepae/pandarallel) package.

Moreover, I would love to stop the pyrochem.REE to drop PM if it is already in the dataset. Preparing plots gets more complicated if the already added column is dropped by this column.

## Related Issue
#99 

## Motivation and Context
Apart from solving a performance issue, it also allows to define in a dataset excluded elements for each sample and which samples should be fitted using the LTE effect or not. When integrating lambda fitting into larger pipelines, it is easier to work on a per sample base than splitting the dataset into multiple data frames upfront.

We have many pipelines that use `pyrochem.REE` for selecting the REE pattern before plotting. We love to maintain the PM column in the pattern, if it is already present in the data set.

## Types of changes

- [ x] New feature (non-breaking change which adds functionality)
- [ x] Breaking change (fix or feature that would cause existing functionality to change)
Keeping the PM column if present is a breaking change. Currently, it is always dropped using `pyrochem.REE`

## Checklist:

- [ x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [] I have updated the documentation accordingly. (Eventually, there is some more to add to `pyrochem.REE`. Not sure.)
- [ ] I have added tests to cover my changes.
- [ x] All new and existing tests passed.

I am happy to make further changes to this PR.

### CODE REVIEW

**Review by morganjwilliams — COMMENTED** [review]
Hey @mmuesly, thanks again for the PR! I've had a look through these change, and added some comments here around the dataframe/series accessors. 

For the lambdas side of things, I had a look at a few changes to adapt existing functions to achieve the same result, without adding much code/extra functions:
* The two helper functions can be condensed into one which has a switch for series/dataframes.
* For each of the algorithms/methods, I've made a few small modifications which allow them to be used with series, together with the helper function to wrap everything together. This makes it a bit easier to maintain and in most cases gives fewer places in the docs to go chasing information.

If OK with you I could push them to your branch, or otherwise post some of it here and we can discuss/edit after merging?

After this is merged I'll also add a few tests to pass series to these functions, and down the line perhaps modify the `pyrolite.util.lambdas.calc_lambdas()` function to deal with series too.

File: pyrolite/geochem/__init__.py [hunk 1]
```diff
@@ -44,8 +44,10 @@ def list_elements(self):
         -------
         The list will have the same ordering as the source DataFrame.
         """
-        fltr = self._obj.columns.isin(_common_elements)
-        return self._obj.columns[fltr].tolist()
+        if isinstance(self._obj, pd.Series):
```
[the remark this answers is not in the document]
Comment (mmuesly) [reply pyrolite/geochem/__init__.py]: Fair point, I changed the PR.

File: pyrolite/geochem/__init__.py [hunk 2]
```diff
@@ -76,7 +80,16 @@ def list_REE(self):
         -------
         The returned list will reorder REE based on atomic number.
         """
-        return [i for i in REE() if i in self._obj.columns]
+        if isinstance(self._obj, pd.Series):
+            if "Pm" in self._obj.index:
```
[the remark this answers is not in the document]
Comment (mmuesly) [reply pyrolite/geochem/__init__.py]: IMHO, there is a limit to the amount of filter conditions in a list comprehension before readability gets harmed, but I have no strong opinion here. I have changed it in the PR.

**Comment by mmuesly** [comment]
@morganjwilliams I pushed my proposal for unit tests.

The invocation with apply for the lambda calculation is what is needed to plugin [pandarallel](https://pypi.org/project/pandarallel/) later for performance reason.
I am open for feedback and happy to change anything in this PR.

Sadly, I could not fix the docs generation error. I do not understand what the broken code is supposed to do.

**Comment by morganjwilliams** [comment]
@mmuesly thanks for adding the test suggestions also, I'll have another read through and look to get this in (apologies, I've been traveling and also had some recent big deadlines). Unless there's anything specific around the high-level API, any small tweaks to organization etc I might look to just update on develop subsequently. I'll also chase up the docs-generation related error, it looks like a `numpy` type conversion thing, so should be easily solvable.

**Pull request merged**